Sources/SwiftSyntaxMacroExpansion/MacroExpansion.swift

```diff
@@ -0,0 +1,203 @@
+import SwiftSyntax
+import SwiftSyntaxMacros
+
+public enum MacroRole {
+  case expression
+  case declaration
+  case accessor
+  case memberAttribute
+  case member
+  case peer
+  case conformance
+  case codeItem
+}
+
+/// Simple diagnostic message
+private struct MacroExpansionError: Error, CustomStringConvertible {
+  let description: String
+}
+
+/// Expand `@freestanding(XXX)` macros.
+public func expandFreestandingMacro(
+  definition: Macro.Type,
+  node: FreestandingMacroExpansionSyntax,
+  in context: some MacroExpansionContext
+) -> String? {
+  do {
+    func _expand(node: some FreestandingMacroExpansionSyntax) throws -> String {
+      let expandedSyntax: Syntax
+      switch definition {
+      case let exprMacroDef as ExpressionMacro.Type:
+        expandedSyntax = try Syntax(exprMacroDef.expansion(of: node, in: context))
+
+      case let declMacroDef as DeclarationMacro.Type:
+        let rewritten = try declMacroDef.expansion(of: node, in: context)
+        expandedSyntax = Syntax(
+          CodeBlockItemListSyntax(
+            rewritten.map {
+              CodeBlockItemSyntax(item: .decl($0))
+            }
+          )
+        )
+
+      case let codeItemMacroDef as CodeItemMacro.Type:
+        let rewritten = try codeItemMacroDef.expansion(of: node, in: context)
+        expandedSyntax = Syntax(CodeBlockItemListSyntax(rewritten))
+
+      default:
+        throw MacroExpansionError(description: "macro doesn't conform to required macro role")
+      }
+      return expandedSyntax.formattedExpansion(definition.formatMode)
+    }
+    return try _openExistential(node, do: _expand)
+  } catch {
+    context.addDiagnostics(from: error, node: node)
+    return nil
+  }
+}
+
+/// Expand `@attached(XXX)` macros.
+public func expandAttachedMacro<Context: MacroExpansionContext>(
+  definition: Macro.Type,
+  macroRole: MacroRole,
+  attributeNode: AttributeSyntax,
+  declarationNode: DeclSyntax,
+  parentDeclNode: DeclSyntax?,
+  in context: Context
+) -> [String]? {
+  do {
+    switch (definition, macroRole) {
+    case (let attachedMacro as AccessorMacro.Type, .accessor):
+      let accessors = try attachedMacro.expansion(
+        of: attributeNode,
+        providingAccessorsOf: declarationNode,
+        in: context
+      )
+      return accessors.map {
+        $0.formattedExpansion(definition.formatMode)
+      }
+
+    case (let attachedMacro as MemberAttributeMacro.Type, .memberAttribute):
+      guard
+        let parentDeclGroup = parentDeclNode?.asProtocol(DeclGroupSyntax.self)
+      else {
+        // Compiler error: 'parentDecl' is mandatory for MemberAttributeMacro.
+        throw MacroExpansionError(description: "parent decl group is nil")
```

Comment:
Any reason why you went to string-based error messages instead of enum cases. I personally prefer enums cases because it doesn’t bury the message somewhere in the source code and enum’s you can actually check programmatically if you want.